### PR TITLE
CO: Make Symfony routes working even without URL rewriting

### PR DESCRIPTION
**Pull request by Quetzacoalt91**
| Questions | Answers |
| --- | --- |
| Branch? | develop |
| Description? | When then URL rewriting is not enabled on Apache, the merchant cannot reach the pages working on Symfony. This PR fixes the issue by adding a missing `index.php/` in the URL |
| Type? | Improvement |
| Category? | CO |
| BC breaks? | Nope |
| Deprecations? | Nope |
| How to test? | Disable the URL rewriting on your Apache server, the product and module pages should still be accessible with a modified URL. |

### CODE REVIEW

File: classes/Link.php [hunk 1]
```diff
@@ -494,6 +494,10 @@ public function getAdminLink($controller, $with_token = true, $sfRouteParams = a
 
         $params = $with_token ? array('token' => Tools::getAdminTokenLite($controller)) : array();
 
+        // Even if URL rewriting is not enabled, the page handled by Symfony must work !
+        // For that, we add an 'index.php' in the URL before the route
+        $index = filter_input(INPUT_SERVER, 'HTTP_MOD_REWRITE')?'':'/index.php';
```
Comment (Shudrum) [inline classes/Link.php]: Can you add some spaces between operators and values?

File: classes/Link.php [hunk 2]
```diff
@@ -503,25 +507,25 @@ public function getAdminLink($controller, $with_token = true, $sfRouteParams = a
                 if (!$redirectLegacy) {
                     if (array_key_exists('id_product', $sfRouteParams)) {
                         if (array_key_exists('deleteproduct', $sfRouteParams)) {
-                            return $this->getBaseLink().basename(_PS_ADMIN_DIR_).'/product/unit/delete/' . $sfRouteParams['id_product'];
+                            return $this->getBaseLink().basename(_PS_ADMIN_DIR_).$index.'/product/unit/delete/' . $sfRouteParams['id_product'];
```
Comment (Shudrum) [inline classes/Link.php]: No spaces between the dot operator please :)

**Comment by Shudrum** [comment]
👍

File: classes/Link.php (hunk 2 above)
Comment (Quetzacoalt91) [inline classes/Link.php]: Wololo, I'm not responsible of this ! :stuck_out_tongue:  Done anyway.

File: classes/Link.php (hunk 1 above)
Comment (Quetzacoalt91) [inline classes/Link.php]: Done

**Comment by mickaelandrieu** [comment]
Please wait before merge I need to test something

File: classes/Link.php [hunk 3]
```diff
@@ -503,25 +507,25 @@ public function getAdminLink($controller, $with_token = true, $sfRouteParams = a
                 if (!$redirectLegacy) {
                     if (array_key_exists('id_product', $sfRouteParams)) {
                         if (array_key_exists('deleteproduct', $sfRouteParams)) {
-                            return $this->getBaseLink().basename(_PS_ADMIN_DIR_).'/product/unit/delete/' . $sfRouteParams['id_product'];
+                            return $this->getBaseLink().basename(_PS_ADMIN_DIR_).$index.'/product/unit/delete/'.$sfRouteParams['id_product'];
```
Comment (mickaelandrieu) [inline classes/Link.php]: you can use:

```php
$kernel->getContainer()->get('router')->generate('admin_product_unit_action', array('action' => 'delete', 'id' => 10), true);
```

Pros:
* no more hardcoded urls here
* absolute urls
* index.php

Cons:
* index.php will always be "printed" : is that an issue ?
* Link class now depends on Symfony router

Comment (Quetzacoalt91) [inline classes/Link.php]: I don't think `Link class now depends on Symfony router` is a bad thing.

If one of its routes is updated, we won't have to change anything here if the URLs are not hardcoded anymore

**Comment by Quetzacoalt91** [comment]
Please note this PR has been updated in order to generate URL to Symfony page by using ... the Symfony router.

Regarding
> Cons:
> * index.php will always be "printed" : is that an issue ?

I have updated the `.htaccess` file, which will remove the index.php from the URL and redirect properly the merchant to the cleaner route. :wink:

File: classes/Link.php [hunk 4]
```diff
@@ -494,35 +494,46 @@ public function getAdminLink($controller, $with_token = true, $sfRouteParams = a
 
         $params = $with_token ? array('token' => Tools::getAdminTokenLite($controller)) : array();
 
+        // Even if URL rewriting is not enabled, the page handled by Symfony must work !
+        // For that, we add an 'index.php' in the URL before the route
+        global $kernel; // sf kernel
+        $sfRouter = $kernel->getContainer()->get('router');
```
Comment (mickaelandrieu) [inline classes/Link.php]: what happens if Symfony kernel is not available at this moment ? For instance on installation process ?

Comment (Quetzacoalt91) [inline classes/Link.php]: [Explosion](https://upload.wikimedia.org/wikipedia/commons/7/79/Operation_Upshot-Knothole_-_Badger_001.jpg)

File: classes/Link.php [hunk 5]
```diff
@@ -494,35 +494,47 @@ public function getAdminLink($controller, $with_token = true, $sfRouteParams = a
 
         $params = $with_token ? array('token' => Tools::getAdminTokenLite($controller)) : array();
 
+        // Even if URL rewriting is not enabled, the page handled by Symfony must work !
+        // For that, we add an 'index.php' in the URL before the route
+        global $kernel; // sf kernel
+
```
Comment (mickaelandrieu) [inline classes/Link.php]: may I suggest something like:

```php
if($kernel instanceof Symfony\Component\HttpKernel\HttpKernelInterface) {
    $sfRouter = $kernel->getContainer()->get('router');
}
```

Comment (Shudrum) [inline classes/Link.php]: 👍

Comment (Quetzacoalt91) [inline classes/Link.php]: Here you go

**Pull request merged**